5m5v-bot.js

```diff
@@ -1,140 +1,27 @@
-var util = require('./lib/util')
-var fs = require('fs')
-var yaml = require('js-yaml')
-var request = require('request')
+const fs = require('fs')
+const yaml = require('js-yaml')
+const Twitter = require('./lib/twitter')
+const TweetFilter = require('./lib/filter')
 
-// loead the config or bust
-try {
-    var config = yaml.safeLoad(fs.readFileSync('5m5v-config.yaml', 'utf8'))
-} catch(e) {
-    throw 'Unable to load config file: ' + e
-}
-
-// fail if there are no users
-if(!config.users.length)
-    throw 'Config needs to contain at least one user'
-
-DELAYTIME = config.delaytime || 2 * 60 * 1000
-
-var users = []
-
-// verify that consumer keys are correct
-request.post( {
-        url: 'https://api.twitter.com/oauth2/token',
-        headers: { 'Authorization': 'Basic ' + Buffer.from(config.consumer_key + ':'  + config.consumer_secret).toString('base64') },
-        form: { 'grant_type': 'client_credentials'}
-    }, ( err, res, body ) => {
-        if(err)
-            throw err
+const isDryRun = process.argv[2] === '--dry-run'
 
-        if(res.statusCode==200)
-            setupUsers()
-        else
-            throw 'Unable to verify consumer API keys: ' + JSON.parse(body).errors[0].message
-});
-
-function setupUsers() {
-
-    // create Twitter instances for all users
-    for(let i = 0; i < config.users.length; i++)
-        users.push(new Twitter({
-            consumer_key: config.consumer_key,
-            consumer_secret: config.consumer_secret,
-            access_token: config.users[i].access_token,
-            access_token_secret: config.users[i].access_token_secret,
-        }))
-
-    // verify that all user credentials are correct
-    var userErrors = 0
-    var promises = []
-    for( let i in users )
-        promises.push( users[i].get('account/verify_credentials').catch(e => {
-            var user = i
-            console.error(`Unable to verify user ${user}: ${e}`) 
-            userErrors ++
-        }))
-
-    Promise.all(promises).then(() => {
-        
-        if( userErrors )
-            console.error('Not all users were verified, terminating')
-        else
-            setupStream()
-    
-    })
+let config = {}
 
+try {
+  config = yaml.safeLoad(fs.readFileSync('5m5v-config.yaml', 'utf8'))
+} catch (error) {
+  throw `Unable to load config file: ${error.message}`
 }
 
-function setupStream() {
-    
-    // many-to-one map of users using each filter
-    var filterUsers = {}
-    for(let i = 0; i < config.users.length; i++)
-        for(let j in config.users[i].filters) {
-            let filterName = config.users[i].filters[j]
-            if(filterName in filterUsers)
-                filterUsers[filterName].push(i)
-            else
-                filterUsers[filterName] = [i]
-        }
+const languages = config.users.map(user => user.language)
 
-    // use first user to stream with
-    var T = users[0]
+const twitter = new Twitter(config.users, new TweetFilter(config.exclude, languages))
```

Comment:
This would definitely be nicer from an API's perspective. I thought about it but the way the bot works is by scanning a list of key words (basically "vegan" and its translations) in all languages at the same time, then checking the matches for all languages. If we split it by language, it will mean one search process per language which would lead to more API calls, I think.

Reply:
However your comment made me realize a flaw in the current implementation: words are searched using "AND" (default) rather than "OR" (I tested it with just English which has only one word so didn't notice). I'll need to fix this!